src/libstd/sys_common/backtrace.rs

```diff
@@ -101,59 +103,42 @@ fn filter_frames(frames: &[Frame],
         return (0, 0);
     }
 
-    // We want to filter out frames with some prefixes
-    // from both top and bottom of the call stack.
+    // Frames to remove from the top of the backtrace.
+    //
+    // The raw form is used so that we don't have to demangle the symbol names.
+    // The `a::b::c` form can show up on Windows/MSVC.
     static BAD_PREFIXES_TOP: &'static [&'static str] = &[
-        "_ZN3std3sys3imp9backtrace",
-        "ZN3std3sys3imp9backtrace",
         "std::sys::imp::backtrace",
-        "_ZN3std10sys_common9backtrace",
-        "ZN3std10sys_common9backtrace",
+        "ZN3std3sys3imp9backtrace",
+
         "std::sys_common::backtrace",
-        "_ZN3std9panicking",
-        "ZN3std9panicking",
+        "ZN3std10sys_common9backtrace",
+
         "std::panicking",
-        "_ZN4core9panicking",
-        "ZN4core9panicking",
+        "ZN3std9panicking",
+
         "core::panicking",
-        "_ZN4core6result13unwrap_failed",
-        "ZN4core6result13unwrap_failed",
+        "ZN4core9panicking",
+
         "core::result::unwrap_failed",
+        "ZN4core6result13unwrap_failed",
+
         "rust_begin_unwind",
         "_ZN4drop",
         "mingw_set_invalid_parameter_handler",
     ];
+
+    // All the frames after these symbols will be removed.
+    // See `rt.rs` and `libtest` for their generation.
     static BAD_PREFIXES_BOTTOM: &'static [&'static str] = &[
-        "_ZN3std9panicking",
-        "ZN3std9panicking",
-        "std::panicking",
-        "_ZN3std5panic",
-        "ZN3std5panic",
-        "std::panic",
-        "_ZN4core9panicking",
-        "ZN4core9panicking",
-        "core::panicking",
-        "_ZN3std2rt10lang_start",
-        "ZN3std2rt10lang_start",
-        "std::rt::lang_start",
-        "panic_unwind::__rust_maybe_catch_panic",
-        "__rust_maybe_catch_panic",
-        "_rust_maybe_catch_panic",
-        "__libc_start_main",
-        "__rust_try",
-        "_start",
-        "main",
-        "BaseThreadInitThunk",
-        "RtlInitializeExceptionChain",
-        "__scrt_common_main_seh",
-        "_ZN4drop",
-        "mingw_set_invalid_parameter_handler",
+        "__rust_begin_backtrace_binary",
+        "__rust_begin_backtrace_test",
     ];
 
     let is_good_frame = |frame: Frame, bad_prefixes: &[&str]| {
         resolve_symname(frame, |symname| {
             if let Some(mangled_symbol_name) = symname {
-                if !bad_prefixes.iter().any(|s| mangled_symbol_name.starts_with(s)) {
+                if !bad_prefixes.iter().any(|s| mangled_symbol_name.contains(s)) {
```

Comment:
I know, but that should be the only cases (no `::h49fe0d471b73c20e` trailers).
I do not expect anybody to define functions which contain the patterns the code is matching for, but a strict comparison would help keeping us on the safe side, just in case ;)